File: go/ajarconnector/connector.go
// SPDX-License-Identifier: Apache-2.0

package ajarconnector

import "github.com/promaka/ajar-connectors/go/eventpb"

// Connector is an inbound connector: it turns a source system's native bytes
// into a canonical Event. Implementations parse native (CoT XML, a CSV row, a
// vendor binary frame, ...) and return a fully-populated event — ideally via
// EventBuilder so the canonical invariants hold by construction.
type Connector interface {
	Normalize(native []byte) (*eventpb.Event, error)
}

// OutboundProfile renders a governed Event into a target system's format. Its
// honesty is the ModeledFields / LossyFields declaration: which canonical fields
// survive translation and which are dropped or approximated. Round-trip
// conformance (canonical -> target -> canonical over the modeled fields) checks
// that claim; see the example package.
type OutboundProfile interface {
	// Target is the human-readable name of the target system (e.g. "TAK").
	Target() string
	// Slug is the stable machine slug for this profile (e.g. "tak-cot").
	Slug() string
	// Version is the profile version (semver-ish string).
	Version() string
	// ModeledFields are the canonical fields this profile faithfully carries.
	ModeledFields() []string
	// LossyFields are the canonical fields this profile drops or approximates.
	LossyFields() []string
	// Render renders the event into the target format's bytes.
	Render(event *eventpb.Event) []byte
}

File: rust/ajar-connector/src/connector.rs
// SPDX-License-Identifier: Apache-2.0
//! Connector traits: inbound normalization and outbound rendering.

use crate::event::Event;

/// An **inbound** connector: turns a source system's native bytes into a
/// canonical [`Event`].
///
/// Implementors parse `native` (CoT XML, a CSV row, a vendor binary frame, …)
/// and return a fully-populated event — ideally via [`crate::EventBuilder`] so
/// the canonical invariants hold by construction.
pub trait Connector {
    /// Error returned when native input cannot be normalized.
    type Error;

    /// Normalizes one native message into a canonical event.
    fn normalize(&self, native: &[u8]) -> Result<Event, Self::Error>;
}

/// An **outbound** connector profile: renders a governed [`Event`] into a
/// target system's format.
///
/// The honesty of an outbound connector is its `modeled_fields` /
/// `lossy_fields` declaration: it states which canonical fields survive the
/// translation and which are dropped or approximated. Round-trip conformance
/// (canonical → target → canonical over the modeled fields) is how that claim
/// is checked; see the example crate for a concrete round-trip test.
pub trait OutboundProfile {
    /// Human-readable name of the target system (e.g. `"TAK"`).
    fn target(&self) -> &str;

    /// Stable machine slug for this profile (e.g. `"tak-cot"`).
    fn slug(&self) -> &str;

    /// Profile version (semver-ish string).
    fn version(&self) -> &str;

    /// Canonical fields this profile faithfully carries into the target.
    fn modeled_fields(&self) -> &[&str];

    /// Canonical fields this profile drops or approximates.
    fn lossy_fields(&self) -> &[&str];

    /// Renders the event into the target format's bytes.
    fn render(&self, event: &Event) -> Vec<u8>;
}
